### ISSUE TITLE
Access to port 10250 (kubelet) for logs: must be public open port ?

### ISSUE THREAD

**Issue by kartoch** [ovh/terraform-ovh-publiccloud-k8s#26]
I made a copy of `/etc/kubernetes/admin.conf` in the fresh cluster to `~/.kube/config`.

When accessing the logs of one of my pods:
```
$ kubectl logs kube-dns-686b776777-qrbmx kubedns --namespace=kube-system
Error from server: Get https://W.X.Y.Z:10250/containerLogs/kube-system/kube-dns-686b776777-qrbmx/kubedns: dial tcp W.X.Y.Z:10250: getsockopt: connection refused
```

The same problem appears when I'm trying to install helm/tiller with or without RBAC support, helm returns the same error as it tried to connect to the same kubelet port.

My previous cluster (using `kubeadm`) was able to access the logs using this way. It seems my `kubectl` client requests access to the port of one of the kubelet instances. Is that the normal behavior ?

**Comment by kartoch** [comment]
My progress so far:

1.  I've open all ports on openstack using security ports => failed

2. I've logged in the node using ssh (address: W.X.Y.Z is public) and ask for logs again:

core@k8s-2 ~ $ sudo /opt/k8s/bin/kubectl --kubeconfig=/etc/kubernetes/admin.conf logs kube-dns-686b776777-qrbmx kubedns --namespace=kube-system 
Error from server: Get https://W.X.Y.Z:10250/containerLogs/kube-system/kube-dns-686b776777-qrbmx/kubedns: dial tcp W.X.Y.Z:10250: getsockopt: connection refused

Any thoughts ?

**Comment by kartoch** [comment]
If i look to the open port on the W.X.Y.Z node:

```
core@k8s-2 ~ $ sudo netstat -tulpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 127.0.0.1:4194          0.0.0.0:*               LISTEN      19198/kubelet       
tcp        0      0 127.0.0.1:10248         0.0.0.0:*               LISTEN      19198/kubelet       
tcp        0      0 127.0.0.1:10249         0.0.0.0:*               LISTEN      2289/kube-proxy     
tcp        0      0 127.0.0.1:10250         0.0.0.0:*               LISTEN      19198/kubelet       
tcp        0      0 127.0.0.1:10251         0.0.0.0:*               LISTEN      1909/kube-scheduler 
tcp        0      0 127.0.0.1:10252         0.0.0.0:*               LISTEN      1875/kube-controlle 
tcp        0      0 127.0.0.1:10255         0.0.0.0:*               LISTEN      19198/kubelet       
tcp6       0      0 :::9099                 :::*                    LISTEN      2978/calico-felix   
tcp6       0      0 :::6443                 :::*                    LISTEN      1900/kube-apiserver 
tcp6       0      0 :::2379                 :::*                    LISTEN      1410/etcd           
tcp6       0      0 :::2380                 :::*                    LISTEN      1410/etcd           
tcp6       0      0 :::10256                :::*                    LISTEN      2289/kube-proxy     
tcp6       0      0 :::22                   :::*                    LISTEN      1/systemd           
udp        0      0 145.239.159.139:68      0.0.0.0:*                           727/systemd-network 
udp        0      0 0.0.0.0:8472            0.0.0.0:*                           -                   
udp6       0      0 fe80::f816:3eff:fe6:546 :::*                                727/systemd-network 

```

The 10250/tcp port is binded to loopback interface, so It cannot be accessed from the public address. So why kubectl tried to access using the public ip address ?!

I didn't saw this problem using my previous clusters, something specific to this deploy script ?

**Comment by yanndegat** [comment]
hi @kartoch 

thanks again for reporting this
there maybe an error in a kubelet.service which spefically binds to 127.0.0.1
i've set it back to the default 0.0.0.0 value which should solve this issue

you may want to try my PR #31 to see if it solves the issue

to access the cluster from outside, you shall add your IP to the security group of the master nodes


we're still in active development for this module so you may encounter a lot of other issues
 
plus most of devs will be off for the next week so there wont be much activity on this module
until mid march

**Comment by kartoch** [comment]
Trying on a flesh cluster 3 nodes public:

```
$ ssh core@54.37.149.97
The authenticity of host '54.37.149.97 (54.37.149.97)' can't be established.
ECDSA key fingerprint is SHA256:oRsBYSK4eUvbkOHlPMwRhywL8hz5+Pgzo2oUsq/JX2E.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '54.37.149.97' (ECDSA) to the list of known hosts.
Last login: Wed Feb 28 16:04:32 UTC 2018 from 51.254.223.129 on pts/2
Container Linux by CoreOS stable (1576.5.0)
core@myk8s-0 ~ $ sudo /opt/k8s/bin/kubectl --kubeconfig /etc/kubernetes/admin.conf get nodes
NAME      STATUS    ROLES     AGE       VERSION
myk8s-0   Ready     master    13m       v1.9.2
myk8s-1   Ready     master    14m       v1.9.2
myk8s-2   Ready     master    13m       v1.9.2
core@myk8s-0 ~ $ sudo /opt/k8s/bin/kubectl --kubeconfig /etc/kubernetes/admin.conf logs     
error: expected 'logs (POD | TYPE/NAME) [CONTAINER_NAME]'.
POD or TYPE/NAME is a required argument for the logs command
See 'kubectl logs -h' for help and examples.
core@myk8s-0 ~ $ sudo /opt/k8s/bin/kubectl --kubeconfig /etc/kubernetes/admin.conf get pods --all-namespaces
NAMESPACE     NAME                              READY     STATUS    RESTARTS   AGE
kube-system   canal-rppwj                       3/3       Running   0          13m
kube-system   canal-rs2ps                       3/3       Running   1          13m
kube-system   canal-wnbdr                       3/3       Running   1          13m
kube-system   kube-apiserver-myk8s-0            1/1       Running   0          12m
kube-system   kube-apiserver-myk8s-1            1/1       Running   0          13m
kube-system   kube-apiserver-myk8s-2            1/1       Running   0          12m
kube-system   kube-controller-manager-myk8s-0   1/1       Running   0          12m
kube-system   kube-controller-manager-myk8s-1   1/1       Running   0          13m
kube-system   kube-controller-manager-myk8s-2   1/1       Running   0          12m
kube-system   kube-dns-686b776777-frdzw         3/3       Running   0          14m
kube-system   kube-proxy-khvcw                  1/1       Running   0          13m
kube-system   kube-proxy-sqgb8                  1/1       Running   0          13m
kube-system   kube-proxy-vq2h5                  1/1       Running   0          13m
kube-system   kube-scheduler-myk8s-0            1/1       Running   0          12m
kube-system   kube-scheduler-myk8s-1            1/1       Running   0          13m
kube-system   kube-scheduler-myk8s-2            1/1       Running   0          12m
core@myk8s-0 ~ $ sudo /opt/k8s/bin/kubectl --kubeconfig /etc/kubernetes/admin.conf logs kube-dns-686b776777-frdzw
Error from server (NotFound): pods "kube-dns-686b776777-frdzw" not found
core@myk8s-0 ~ $ sudo /opt/k8s/bin/kubectl --kubeconfig /etc/kubernetes/admin.conf logs kube-dns-686b776777-frdzw --namespace=kube-system
Error from server (BadRequest): a container name must be specified for pod kube-dns-686b776777-frdzw, choose one of: [kubedns dnsmasq sidecar]
core@myk8s-0 ~ $ sudo /opt/k8s/bin/kubectl --kubeconfig /etc/kubernetes/admin.conf logs kube-dns-686b776777-frdzw dnsmasq --namespace=kube-system
Error from server: Get https://54.37.149.89:10250/containerLogs/kube-system/kube-dns-686b776777-frdzw/dnsmasq: dial tcp 54.37.149.89:10250: getsockopt: connection refused
```

**Comment by yanndegat** [comment]
currently making tests, will keep you informed

-yann

**Comment by kartoch** [comment]
It works using e3b654b257f801f5695245a287ea88b0a99f063b which include your PR #31 :
```
core@k8s-0 ~ $  sudo /opt/k8s/bin/kubectl --kubeconfig /etc/kubernetes/admin.conf get pod --namespace=kube-system
NAME                            READY     STATUS    RESTARTS   AGE
canal-65q86                     3/3       Running   0          11m
canal-8z6xz                     3/3       Running   0          11m
canal-bnwmw                     3/3       Running   1          9m
canal-njpx5                     3/3       Running   1          10m
canal-nlj89                     3/3       Running   1          10m
kube-apiserver-k8s-0            1/1       Running   0          10m
kube-apiserver-k8s-1            1/1       Running   0          9m
kube-apiserver-k8s-2            1/1       Running   0          9m
kube-apiserver-k8s-3            1/1       Running   0          10m
kube-apiserver-k8s-4            1/1       Running   0          8m
kube-controller-manager-k8s-0   1/1       Running   0          10m
kube-controller-manager-k8s-1   1/1       Running   0          9m
kube-controller-manager-k8s-2   1/1       Running   0          9m
kube-controller-manager-k8s-3   1/1       Running   0          10m
kube-controller-manager-k8s-4   1/1       Running   0          8m
kube-dns-686b776777-kb5rm       3/3       Running   0          11m
kube-proxy-dxxzg                1/1       Running   0          11m
kube-proxy-nmc7f                1/1       Running   0          10m
kube-proxy-nt4tx                1/1       Running   0          11m
kube-proxy-vj82d                1/1       Running   0          10m
kube-proxy-wz9p2                1/1       Running   0          9m
kube-scheduler-k8s-0            1/1       Running   0          10m
kube-scheduler-k8s-1            1/1       Running   0          9m
kube-scheduler-k8s-2            1/1       Running   0          9m
kube-scheduler-k8s-3            1/1       Running   0          10m
kube-scheduler-k8s-4            1/1       Running   0          8m
core@k8s-0 ~ $ sudo /opt/k8s/bin/kubectl --kubeconfig /etc/kubernetes/admin.conf logs kube-dns-686b776777-kb5rm dnsmasq --namespace=kube-system
I0228 17:56:08.792476       1 main.go:76] opts: {{/usr/sbin/dnsmasq [-k --cache-size=1000 --no-negcache --log-facility=- --server=/dc1.local/127.0.0.1#10053 --server=/in-addr.arpa/127.0.0.1#10053 --server=/ip6.arpa/127.0.0.1#10053] true} /etc/k8s/dns/dnsmasq-nanny 10000000000}
I0228 17:56:08.792625       1 nanny.go:94] Starting dnsmasq [-k --cache-size=1000 --no-negcache --log-facility=- --server=/dc1.local/127.0.0.1#10053 --server=/in-addr.arpa/127.0.0.1#10053 --server=/ip6.arpa/127.0.0.1#10053]
I0228 17:56:09.161945       1 nanny.go:119] 
W0228 17:56:09.162070       1 nanny.go:120] Got EOF from stdout
I0228 17:56:09.162527       1 nanny.go:116] dnsmasq[12]: started, version 2.78 cachesize 1000
I0228 17:56:09.162547       1 nanny.go:116] dnsmasq[12]: compile time options: IPv6 GNU-getopt no-DBus no-i18n no-IDN DHCP DHCPv6 no-Lua TFTP no-conntrack ipset auth no-DNSSEC loop-detect inotify
I0228 17:56:09.162589       1 nanny.go:116] dnsmasq[12]: using nameserver 127.0.0.1#10053 for domain ip6.arpa 
I0228 17:56:09.162599       1 nanny.go:116] dnsmasq[12]: using nameserver 127.0.0.1#10053 for domain in-addr.arpa 
I0228 17:56:09.162603       1 nanny.go:116] dnsmasq[12]: using nameserver 127.0.0.1#10053 for domain dc1.local 
I0228 17:56:09.162755       1 nanny.go:116] dnsmasq[12]: reading /etc/resolv.conf
I0228 17:56:09.162821       1 nanny.go:116] dnsmasq[12]: using nameserver 127.0.0.1#10053 for domain ip6.arpa 
I0228 17:56:09.162830       1 nanny.go:116] dnsmasq[12]: using nameserver 127.0.0.1#10053 for domain in-addr.arpa 
I0228 17:56:09.162834       1 nanny.go:116] dnsmasq[12]: using nameserver 127.0.0.1#10053 for domain dc1.local 
I0228 17:56:09.162893       1 nanny.go:116] dnsmasq[12]: using nameserver 213.186.33.99#53
I0228 17:56:09.163705       1 nanny.go:116] dnsmasq[12]: read /etc/hosts - 7 addresses
```

**Comment by yanndegat** [comment]
merged a fix on master. should be working now.

tested on both examples

fixed by #33

**Comment by jayant12489** [comment]
Hi Team Steel issue has been same please provde or update solution 
[root@jayant kubernetes]# kubectl get pod
NAME               READY   STATUS    RESTARTS   AGE
myfirstpod1        1/1     Running   1          7h4m
opencart-testing   1/1     Running   0          21m
[root@jayant kubernetes]# kubectl exec opencart-testing env
Error from server: error dialing backend: dial tcp XX.XX.XX.XX:10250: connect: connection refused
